server/rpc.go

```diff
@@ -31,31 +36,56 @@ func (srv *Server) rpcHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	ws, err := upgrader.Upgrade(w, r, nil)
-
 	if err != nil {
 		log.Println(errors.Wrap(err, "websocket upgrade failed"))
 		return
 	}
-	defer ws.Close()
+
+	log.Printf("Opened ws with IP: %s, %s User-Agent: %s", clientIP(r), r.RemoteAddr, r.Header.Get("User-Agent"))
+	defer func() {
+		log.Printf("Closed ws with IP: %s, %s User-Agent: %s", clientIP(r), r.RemoteAddr, r.Header.Get("User-Agent"))
+		ws.Close()
+	}()
+
 	ws.SetReadLimit(websocketReadLimit)
 
 	sendCh := make(chan interface{})
 	defer close(sendCh)
 
 	go func() {
+		isClosed := false
 		for {
 			res, ok := <-sendCh
 			if !ok {
+				log.Printf("channel closed. IP: %s, %s User-Agent: %s", clientIP(r), r.RemoteAddr, r.Header.Get("User-Agent"))
 				return
 			}
+
+			if isClosed {
+				log.Printf("received a message %v after ws closed. IP: %s, %s User-Agent: %s", res, clientIP(r), r.RemoteAddr, r.Header.Get("User-Agent"))
+				continue
+			}
+
 			if v, ok := res.(rune); ok && v == 'h' {
-				ws.WriteMessage(websocket.TextMessage, []byte("h"))
+				if srv.writeDeadline > 0 {
+					ws.SetWriteDeadline(time.Now().Add(srv.writeDeadline))
+				}
+
+				if err := ws.WriteMessage(websocket.TextMessage, []byte("h")); err != nil {
+					log.Println(errors.Wrap(err, "websocket write heartbeat"))
+				}
+
 				continue
 			}
+
+			if srv.writeDeadline > 0 {
+				ws.SetWriteDeadline(time.Now().Add(srv.writeDeadline))
+			}
+
 			if err := ws.WriteJSON(res); err != nil {
 				log.Println(errors.Wrap(err, "websocket write failed"))
 				ws.Close()
-				return
+				isClosed = true
```

Comment:
any reason we're setting the flag?Why not just log here and exit the goroutine? instead of having it to continue to for-loop?

Reply:
Also, I'm not sure this is fixing anything? Isn't it going to hang the goroutine?

Reply:
ah if we just return here, `sendCh` will stay opened for reading and it will make https://github.com/walletlink/walletlink/blob/master/server/rpc/pub_sub.go#L143 blocked.

So the solution is reading messages from the channel until it get closed.